libraries/AP_Baro/AP_Baro.cpp

```diff
@@ -611,83 +706,110 @@ void AP_Baro::init(void)
+        if (!probe_spi_dev(AP_Baro_MS5611::probe, HAL_BARO_MS5611_SPI_IMU_NAME)) {
+            return;
+        }
 #endif
 #endif  // AP_BARO_MS5611_ENABLED
         break;
 
     case AP_BoardConfig::PX4_BOARD_PCNC1:
 #if AP_BARO_ICM20789_ENABLED
-        ADD_BACKEND(AP_Baro_ICM20789::probe(*this,
-                                            std::move(GET_I2C_DEVICE(1, 0x63)),
-                                            std::move(hal.spi->get_device(HAL_INS_MPU60x0_NAME))));
+        if (!probe_icm20789(1, 0x63, HAL_INS_MPU60x0_NAME)) {
+            return;
+        }
 #endif
         break;
 
     default:
         break;
     }
 #elif HAL_BARO_DEFAULT == HAL_BARO_LPS25H_IMU_I2C
-	ADD_BACKEND(AP_Baro_LPS2XH::probe_InvensenseIMU(*this,
-                                                    std::move(GET_I2C_DEVICE(HAL_BARO_LPS25H_I2C_BUS, HAL_BARO_LPS25H_I2C_ADDR)),
-                                                    HAL_BARO_LPS25H_I2C_IMU_ADDR));
+    {
+        const auto *dev = GET_I2C_DEVICE(HAL_BARO_LPS25H_I2C_BUS, HAL_BARO_LPS25H_I2C_ADDR);
+        if (dev) {
+            auto *backend = AP_Baro_LPS2XH::probe_InvensenseIMU(*this,
+                                                                dev,
+                                                                HAL_BARO_LPS25H_I2C_IMU_ADDR);
+            if (backend != nullptr) {
+                ADD_BACKEND(backend);
+            } else {
+                delete dev;
+            }
+        }
+    }
 #elif HAL_BARO_DEFAULT == HAL_BARO_20789_I2C_I2C
-    ADD_BACKEND(AP_Baro_ICM20789::probe(*this,
-                                        std::move(GET_I2C_DEVICE(HAL_BARO_20789_I2C_BUS, HAL_BARO_20789_I2C_ADDR_PRESS)),
-                                        std::move(GET_I2C_DEVICE(HAL_BARO_20789_I2C_BUS, HAL_BARO_20789_I2C_ADDR_ICM))));
+    // probe_icm20789 returns false if there's no space for more backends
+    if (!probe_icm20789(HAL_BARO_20789_I2C_BUS, HAL_BARO_20789_I2C_ADDR_PRESS, HAL_BARO_20789_I2C_BUS, HAL_BARO_20789_I2C_ADDR_ICM)) {
+        return;
+    }
 #elif HAL_BARO_DEFAULT == HAL_BARO_20789_I2C_SPI
-    ADD_BACKEND(AP_Baro_ICM20789::probe(*this,
-                                        std::move(GET_I2C_DEVICE(HAL_BARO_20789_I2C_BUS, HAL_BARO_20789_I2C_ADDR_PRESS)),
-                                        std::move(hal.spi->get_device("icm20789"))));
+    // probe_icm20789 returns false if there's no space for more backends
+    if (!probe_icm20789(HAL_BARO_20789_I2C_BUS, HAL_BARO_20789_I2C_ADDR_PRESS, "icm20789")) {
+        return;
+    }
 #endif
 
     // can optionally have baro on I2C too
     if (_ext_bus >= 0) {
 #if APM_BUILD_TYPE(APM_BUILD_ArduSub)
 #if AP_BARO_MS5837_ENABLED
-        ADD_BACKEND(AP_Baro_MS5837::probe(*this,
-                                          std::move(GET_I2C_DEVICE(_ext_bus, HAL_BARO_MS5837_I2C_ADDR))));
+        if (!probe_i2c_dev(AP_Baro_MS5837::probe, _ext_bus, HAL_BARO_MS5837_I2C_ADDR)) {
```

Comment:
Change to make these functions void (and remove the comments about what they do!) and swipe CHECK_UNREG_RETURN from AP_Compass library

`RETURN_IF_NO_SPACE;` ?

Reply:
Done